src/rgw/rgw_data_sync.cc

```diff
@@ -1413,6 +1413,179 @@ class RGWDataSyncSingleEntryCR : public RGWCoroutine {
   }
 };
 
+class RGWDataIncrementalSyncFullObligationCR: public RGWCoroutine {
+  RGWDataSyncCtx *sc;
+  RGWDataSyncEnv *sync_env;
+  rgw_bucket_shard source_bs;
+  std::string error_marker;
+  ceph::real_time timestamp;
+  RGWSyncTraceNodeRef tn;
+  rgw_bucket_index_marker_info remote_info;
+  rgw_pool pool;
+  rgw_raw_obj error_repo;
+  uint32_t sid;
+  rgw_bucket_shard bs;
+  std::vector<store_gen_shards>::const_iterator each;
+
+  rgw_raw_obj datalog_oid_for_error_repo(rgw_bucket_shard& bs) {
+    int datalog_shard = sync_env->store->svc()->datalog_rados->choose_oid(bs);
+    string oid = RGWDataSyncStatusManager::shard_obj_name(sc->source_zone, datalog_shard);
+    return rgw_raw_obj(pool, oid + ".retry");
+  }
+
+public:
+  RGWDataIncrementalSyncFullObligationCR(RGWDataSyncCtx *_sc, rgw_bucket_shard& _source_bs, std::string _error_marker,
+                         ceph::real_time& _timestamp, RGWSyncTraceNodeRef& _tn)
+    : RGWCoroutine(_sc->cct), sc(_sc), sync_env(_sc->env),
+      source_bs(_source_bs), error_marker(_error_marker), timestamp(_timestamp), tn(_tn) {
+        tn = sync_env->sync_tracer->add_node(_tn, "error_repo", SSTR(bucket_shard_str(source_bs)));
+      }
+
+  int operate(const DoutPrefixProvider *dpp) override {
+    reenter(this) {
+      yield call(new RGWReadRemoteBucketIndexLogInfoCR(sc, source_bs.bucket, &remote_info));
+      if (retcode < 0) {
+        return set_cr_error(retcode);
+      }
+
+      each = remote_info.generations.cbegin();
+      for (; each != remote_info.generations.cend(); each++) {
+        for (sid = 0; sid < each->num_shards; sid++) {
+          bs.bucket = source_bs.bucket;
+          bs.shard_id = sid;
+          error_repo = datalog_oid_for_error_repo(bs);
+          tn->log(10, SSTR("writing shard_id " << sid << " of gen " << each->gen << " to error repo for retry"));
+          yield_spawn_window(rgw::error_repo::write_cr(sync_env->store->svc()->rados, error_repo,
+                            rgw::error_repo::encode_key(bs, each->gen),
+                            timestamp), cct->_conf->rgw_data_sync_spawn_window,
+                            [&](uint64_t stack_id, int ret) {
+                              if (ret < 0) {
+                                retcode = ret;
+                              }
+                              return 0;
+                            });
```

Comment:
since we're starting to repeat this callback a bit, what do you think about a helper function?
```c++
// returns a lambda that captures negative error codes into the given int&
auto capture_error(int& retcode) {
  return [&retcode] (uint64_t stack_id, int ret) {
    if (ret < 0) {
      retcode = ret;
    }
    return ret;
  };
}
```
example use:
```c++
yield_spawn_window(rgw::error_repo::write_cr(...),
                   cct->_conf->rgw_data_sync_spawn_window,
                   capture_error(retcode));
```